magento/magento2
# Don't assume a frontend cache section exists in the env.php file.

### Description (*)
Code exists in Magento that assumes your `app/etc/env.php` will always contain a `frontend` section if there is a `cache` section. Like this:
```php
    'cache' => [
        'frontend' => [
            ...
        ]
     ]
```

But that's not always the case. All shops that were setup before Magento 2.3.1 didn't have a `cache` section in their `env.php` file in a default installation (if they didn't use Redis as caching).
If they now upgrade to Magento 2.4.4 or higher and use a graphql endpoint, Magento [automatically writes](https://github.com/magento/magento2/blob/2.4.6-p3/app/code/Magento/GraphQlCache/Model/CacheId/CacheIdCalculator.php#L124-L128) `cache > graphql > id_salt` to the `env.php` file and so introduces a `cache` section without a `frontend` section.

This then leads to this crash on every execution within Magento (frontend, backoffice, `bin/magento`, ...):
```
Warning: Undefined array key "frontend" in lib/internal/Magento/Framework/App/Cache/Frontend/Pool.php on line 90
```

This PR fixes this.

### Related Pull Requests

### Fixed Issues (if relevant)
1. Fixes #35812
2. Fixes #35861

### Manual testing scenarios (*)
1. Fully setup a working Magento
2. Edit the `app/etc/env.php` file and remove the entire `frontend` section in the `cache` section
3. Run `bin/magento` or visit the frontend
4. Expected is to not see an error/crash
5. Run added unit test with and without fix. Without the fix, the test will fail (`vendor/bin/phpunit -c dev/tests/unit/phpunit.xml.dist lib/internal/Magento/Framework/App/Test/Unit/Cache/Frontend/PoolTest.php`)

More elaborate testing scenario that can cause this situation:
1. Using PHP 7.2 and composer v1, setup Magento 2.3.0 and install it fully (don't configure it with Redis caching)
2. Double check that the `app/etc/env.php` file contains no `cache` section
3. Using PHP 8.2 and composer v2, upgrade this installation to Magento 2.4.6-p3 (or 2.4-develop)
4. Run `bin/magento setup:upgrade`
6. Your `app/etc/env.php` file will be updated to a new format, but will still not contain a `cache` section.
7. Confirm that Magento works correctly
8. On the frontend visit the `/graphql` url
9. Check your `app/etc/env.php` file and see that a `cache` section has appeared with a `graphql` section but no `frontend` section
10. Visit the homepage or run `bin/magento`
11. See errors/crashes

### Questions or comments
Should we expect that when we run `bin/magento setup:upgrade` it will update your `app/etc/env.php` file with all missing sections? Would people expect this? Or will this cause issues in some scenario's? Running `bin:magento setup:config:set` without extra flags does exactly that. That seems to be some kind of hidden feature ...

### Contribution checklist (*)


### Resolved issues:
1. [x] resolves magento/magento2#38363: Don't assume a frontend cache section exists in the env.php file.
